Source/WebCore/dom/Document.cpp

```diff
@@ -1330,6 +1330,54 @@ const Color& Document::themeColor()
     return m_cachedThemeColor;
 }
 
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+const std::optional<SpatialBackdropSource>& Document::spatialBackdropSource()
+{
+    return m_cachedSpatialBackdropSource;
+}
+
+void Document::spatialBackdropLinkElementChanged()
+{
+    spatialBackdropSourceChanged();
+}
+
+void Document::spatialBackdropSourceChanged()
+{
+    if (!settings().webPageSpatialBackdropEnabled())
+        return;
+
+    auto newSpatialBackdropSource = determineActiveSpatialBackdropSource();
+    auto oldSpatialBackdropSource = m_cachedSpatialBackdropSource;
+    m_cachedSpatialBackdropSource = newSpatialBackdropSource;
+
+    if (newSpatialBackdropSource == oldSpatialBackdropSource)
+        return;
+
+    if (RefPtr page = this->page())
+        page->chrome().client().spatialBackdropSourceChanged();
+}
+
+std::optional<SpatialBackdropSource> Document::determineActiveSpatialBackdropSource()
+{
+    auto sourceURL = m_url.url();
+    if (!sourceURL.isValid())
+        return std::nullopt;
+
+    for (auto& linkElement : descendantsOfType<HTMLLinkElement>(*this)) {
+        if (!linkElement.isSpatialBackdrop())
+            continue;
+
+        auto modelURL = linkElement.href();
+        if (!modelURL.isValid())
+            continue;
```

Comment:
I think a standard for a feature like this would just pick the first qualifying `link` element and then if it doesn't work give up. Or it would actually fully try each of them. But not only validate a single condition but not any of the others.

Reply:
Attempting the load so we end up firing an `error` event would probably be more useful.